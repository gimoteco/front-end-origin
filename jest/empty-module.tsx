import * as React from 'react';
const SvgrMock = () => <span />;

export default SvgrMock;

export const ReactComponent = SvgrMock;
