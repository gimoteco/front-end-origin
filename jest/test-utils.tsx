import { render } from '@testing-library/react';
import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import theme from '../src/styles/theme';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
