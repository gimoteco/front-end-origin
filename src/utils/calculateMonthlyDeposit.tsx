import {
  addDays,
  differenceInMonths,
  endOfMonth,
  startOfMonth
} from 'date-fns';

export function calculateMonthlyDeposit(total: number, reachDate: Date) {
  const thisMonth = startOfMonth(new Date());
  const endMonth = addDays(endOfMonth(reachDate), 1);
  const numberOfDeposits = differenceInMonths(endMonth, thisMonth);
  const depositValue = total / numberOfDeposits;
  return { numberOfDeposits, depositValue };
}
