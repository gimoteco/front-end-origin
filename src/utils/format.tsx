import { differenceInMonths } from 'date-fns';

const formatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'USD',
  minimumFractionDigits: 0
});

export function formatMoney(value: number) {
  return formatter.format(value);
}

export function calculateMonthlyDeposit(total: number, reachDate: Date) {
  const today = new Date();
  const numberOfDeposits = differenceInMonths(reachDate, today);
  const depositValue = total / numberOfDeposits;
  return { numberOfDeposits, depositValue };
}
