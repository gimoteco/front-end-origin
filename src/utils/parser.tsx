import { format as formatDate } from 'date-fns';

export function parseMoney(currency: string) {
  return Number(currency.replace(/[^0-9.-]+/g, ''));
}

export function parseDateParts(date: Date) {
  const month = formatDate(date, 'LLLL');
  const year = formatDate(date, 'yyyy');
  return { month, year };
}
