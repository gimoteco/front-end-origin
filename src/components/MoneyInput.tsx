import * as React from 'react';
import { MaskedInputProps } from 'react-text-mask';
import createNumberMask from 'text-mask-addons/dist/createNumberMask';
import { parseMoney } from '../utils/parser';
import * as Style from './MoneyInput.styles';

const defaultMaskOptions = {
  prefix: '',
  suffix: '',
  includeThousandsSeparator: true,
  thousandsSeparatorSymbol: ',',
  allowDecimal: true,
  decimalSymbol: '.',
  decimalLimit: 2,
  integerLimit: 10,
  allowNegative: false,
  allowLeadingZeroes: false
};

const currencyMask = createNumberMask(defaultMaskOptions);
interface MoneyInputProps extends Omit<MaskedInputProps, 'onChange'> {
  onChange(newValue: number | null): void;
}

function MoneyInput({ onChange, ...props }: MoneyInputProps) {
  return (
    <Style.InputWrapper>
      <Style.Prefix htmlFor={props.id}>$</Style.Prefix>
      <Style.MaskedTextInput
        mask={currencyMask}
        {...props}
        onChange={e =>
          onChange(e.target.value ? parseMoney(e.target.value) : null)
        }
      />
    </Style.InputWrapper>
  );
}

export default React.memo(MoneyInput);
