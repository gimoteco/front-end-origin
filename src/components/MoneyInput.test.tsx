import * as React from 'react';
import { fireEvent, render } from '../../jest/test-utils';
import { formatMoney } from '../utils/format';
import MoneyInput from './MoneyInput';

describe('MoneyInput', () => {
  it('should format the value', async () => {
    const value = 3000;
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <MoneyInput id="moneyInput" onChange={onChangeMock} />
    );
    const input = (await getByTestId('moneyInput')) as HTMLInputElement;

    fireEvent.change(input, {
      target: { value }
    });

    expect(input.value).toBe(formatMoney(value).replace('$', ''));
    expect(onChangeMock).toHaveBeenCalledWith(value);
  });
});
