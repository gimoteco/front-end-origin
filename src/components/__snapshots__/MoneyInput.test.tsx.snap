// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneyInput should render according snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa djcBRL"
  >
    <label
      class="sc-htpNat eRwJqC"
      for="some-id"
    >
      $
    </label>
    <input
      class="sc-bwzfXH cvgptw"
      data-testid="some-id"
      id="some-id"
      value="40000"
    />
  </div>
</div>
`;
